src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/SpanNDHelpers.cs

```diff
@@ -0,0 +1,508 @@
+            }
+
+            // Store any remaining values one-by-one
+            if (remainder > 1)
+            {
+                ReverseInner(ref Unsafe.Add(ref buf, offset), (nuint)remainder);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static unsafe void Reverse<T>(ref T elements, nuint length)
+        {
+            Debug.Assert(length > 1);
+
+            if (!RuntimeHelpers.IsReferenceOrContainsReferences<T>())
+            {
+                if (sizeof(T) == sizeof(byte))
+                {
+                    Reverse(ref Unsafe.As<T, byte>(ref elements), length);
+                    return;
+                }
+                else if (sizeof(T) == sizeof(char))
+                {
+                    Reverse(ref Unsafe.As<T, char>(ref elements), length);
+                    return;
+                }
+                else if (sizeof(T) == sizeof(int))
+                {
+                    Reverse(ref Unsafe.As<T, int>(ref elements), length);
+                    return;
+                }
+                else if (sizeof(T) == sizeof(long))
+                {
+                    Reverse(ref Unsafe.As<T, long>(ref elements), length);
+                    return;
+                }
+            }
+
+            ReverseInner(ref elements, length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void ReverseInner<T>(ref T elements, nuint length)
+        {
+            Debug.Assert(length > 1);
+
+            ref T first = ref elements;
+            ref T last = ref Unsafe.Subtract(ref Unsafe.Add(ref first, length), 1);
+            do
+            {
+                T temp = first;
+                first = last;
+                last = temp;
+                first = ref Unsafe.Add(ref first, 1);
+                last = ref Unsafe.Subtract(ref last, 1);
+            } while (Unsafe.IsAddressLessThan(ref first, ref last));
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static nint CalculateTotalLength(ref nint[] lengths)
+        {
+            nint totalLength = 1;
+            for (int i = 0; i < lengths.Length; i++)
+            {
+                totalLength *= lengths[i];
+            }
+
+            return totalLength;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static nint CalculateTotalLength(ReadOnlySpan<nint> lengths)
+        {
+            nint totalLength = 1;
+            for (int i = 0; i < lengths.Length; i++)
+            {
+                totalLength *= lengths[i];
+            }
+
+            return totalLength;
+        }
+
+        /// <summary>
+        /// Gets the set of strides that can be used to calculate the offset of n-dimensions in a 1-dimensional layout
+        /// </summary>
+        /// <returns></returns>
+        public static nint[] CalculateStrides(int rank, ReadOnlySpan<nint> lengths)
```

Comment:
Why is `rank` needed, it should simply be `lengths.Length`, no?

Reply:
Should this instead be taking in a `Span<nint> destination` that is asserted to be the same length as `lengths`?

Reply:
Removed rank as you are right its not needed.

I didn't have it take a span as we need it to be an array for the Tensor and didn't want to have to `ToArray` it everywhere, but if you think thats better I can.

Reply:
I'll make that change.